### PR TITLE
SNOW-956848 Store underlying error in SnowparkSQLException

### PR DESCRIPTION
**Pull request by sfc-gh-bchinn**
1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   https://snowflakecomputing.atlassian.net/browse/SNOW-956848

   This is mostly addressed by https://github.com/snowflakedb/snowpark-python/pull/1111, but I think this is a better long-term solution, for the user to have access to any information in the underlying error.

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

### CODE REVIEW

File: src/snowflake/snowpark/_internal/error_message.py [hunk 1]
```diff
@@ -239,8 +239,8 @@ def PLAN_CREATE_DYNAMIC_TABLE_FROM_DDL_DML_OPERATIONS() -> SnowparkCreateDynamic
     @staticmethod
     def DF_ALIAS_NOT_RECOGNIZED(alias: str) -> SnowparkDataframeException:
         return SnowparkDataframeException(
-            f"DataFrame alias unrecognized. A subset of columns corresponding to Dataframe alias '{alias}' can not be found. "
-            "1208",
+            f"DataFrame alias unrecognized. A subset of columns corresponding to Dataframe alias '{alias}' can not be found. ",
+            error_code="1208",
```
Comment (sfc-gh-bchinn) [inline src/snowflake/snowpark/_internal/error_message.py:255]: 😱 This almost makes me wish Python didn't allow implicit string concatenation

**Review by sfc-gh-aalin — APPROVED** [review]
LGTM, this looks much cleaner, thanks for the changes.

**Comment by sfc-gh-aalin** [comment]
Adding @sfc-gh-aling as a reviewer.

It looks like the two of us had the same idea :)

**Review by sfc-gh-sfan — APPROVED** [review]

**Comment by sfc-gh-bchinn** [comment]
@sfc-gh-yixie @sfc-gh-sfan can one of you merge?

**Comment by sfc-gh-sfan** [comment]
There is a CLA bot thingy that I'll check with the team and see what to do about it.

**Comment by sfc-gh-bchinn** [comment]
@sfc-gh-sfan any updates on this?

File: CHANGELOG.md [hunk 2]
```diff
@@ -6,7 +6,7 @@
 
 - Added support for managing case sensitivity in `DataFrame.to_local_iterator()`.
 - Added support for specifying vectorized UDTF's input column names by using the optional parameter `input_names` in `UDTFRegistration.register/register_file` and `functions.pandas_udtf`. By default, `RelationalGroupedDataFrame.applyInPandas` will infer the column names from current dataframe schema.
-- Add `sql_error_code` and `raw_message` attributes to `SnowflakeSQLException` when it is caused by a SQL exception.
+- Add `sql_error_code` and `raw_message` attributes to `SnowflakeSQLException` when it is caused by a SQL exception. Also add the `conn_error` attribute that stores the whole underlying exception from `snowflake-connector-python`
```
Comment (sfc-gh-aalam) [inline CHANGELOG.md:9]: can you add a new entry for the next release. We shouldn't be updating old changelogs

**Review by sfc-gh-aalam — APPROVED** [review]
Outside of changelog, other changes look good to me

File: CHANGELOG.md (hunk 2 above)
Comment (sfc-gh-bchinn) [reply CHANGELOG.md]: done

**Comment by sfc-gh-bchinn** [comment]
@sfc-gh-aalin @sfc-gh-sfan Can one of you merge? I don't have write access

**Comment by sfc-gh-aalin** [comment]
> @sfc-gh-aalin @sfc-gh-sfan Can one of you merge? I don't have write access

Alas, neither do I

**Pull request merged**